react-front-end/tsrc/modules/ViewerModule.ts

```diff
@@ -248,3 +248,243 @@ export const buildLightboxNavigationHandler = (
+ * @param version Version of the Item which the attachments belong to.
+ * @param getViewerDetails Function called to retrieve viewer detail for each attachment.
+ */
+export const buildAttachmentsAndViewerDefinitions = async (
+  attachments: OEQ.Search.Attachment[],
+  uuid: string,
+  version: number,
+  getViewerDetails: (
+    mimeType: string
+  ) => Promise<OEQ.MimeType.MimeTypeViewerDetail>
+): Promise<E.Either<string[], AttachmentAndViewerDefinition[]>> => {
+  const either: E.Either<string, AttachmentAndViewerDefinition>[] =
+    await Promise.all(
+      pipe(
+        attachments,
+        A.partitionMapWithIndex((index, attachment) =>
+          attachment.brokenAttachment === true
+            ? E.left({ index, attachment })
+            : E.right({
+                index,
+                attachment: updateAttachmentForCustomInfo(attachment),
+              })
+        ),
+        ({ left, right }) => {
+          const brokenAttachmentsAndViewerDefinition = left.map(
+            ({ index, attachment }) => ({
+              viewerDefinition: Promise.resolve(
+                E.right<string, AttachmentAndViewerDefinition>({ attachment })
+              ),
+              index,
+            })
+          );
+          const liveAttachmentsAndViewerDefinition = right.map(
+            ({ index, attachment }) => ({
+              viewerDefinition: buildLiveAttachmentsAndViewerDefinitions(
+                attachment,
+                uuid,
+                version,
+                getViewerDetails
+              ),
+              index,
+            })
+          );
+
+          // Merge the two collections and sort as per the order on the server.
+          return brokenAttachmentsAndViewerDefinition
+            .concat(liveAttachmentsAndViewerDefinition)
+            .sort(
+              ({ index: prevIndex }, { index: nextIndex }) =>
+                prevIndex - nextIndex
+            )
+            .map(({ viewerDefinition }) => viewerDefinition);
+        }
+      )
+    );
+
+  return pipe(either, A.separate, ({ left, right }) =>
+    A.isEmpty(left) ? E.right(right) : E.left(left)
+  );
+};
+
+/**
+ * Build Lightbox entries based on a list of attachments and viewer definitions. Attachments not viewed in the
+ * Lightbox are dropped.
+ *
+ * @param attachmentsAndViewerDefinitions A list of attachments and their viewer definitions.
+ */
+export const buildLightboxEntries: (
+  attachmentsAndViewerDefinitions: AttachmentAndViewerDefinition[]
+) => LightboxEntry[] = flow(
+  A.map(({ attachment: { id, description, mimeType }, viewerDefinition }) =>
+    pipe(
+      viewerDefinition,
+      O.fromNullable,
+      O.filter(([viewer]) => viewer === "lightbox"),
+      O.map(([_, viewerUrl]) => ({
+        src: viewerUrl,
+        title: description,
+        mimeType: mimeType ?? "",
+        id,
+      }))
+    )
+  ),
+  A.filter(O.isSome),
+  A.map((entry) => entry.value)
```

Comment:
I thought I wrote this comment already, but I can't find it. So if you get this twice...

Would this be more idiomatic as:

```suggestion
  A.map(O.getOrElse(absurd))
```

Requires:

```typescript
import {absurd} from "fp-ts/function";
```